### PR TITLE
[Core][Node] Adding const-fcts for GetDof

### PR DESCRIPTION
**Pull request by philbucher**
needed for #2993

### CODE REVIEW

**Review by loumalouomega — DISMISSED** [review]
We can wait for the technical committee, but I assume they will not oppose to add some const methods

**Review by pooyan-dadvand — CHANGES_REQUESTED** [review]
In my opinion we should put the GetDof as const only and use the pAddDof when we want to create if not exist. I will add this to @KratosMultiphysics/technical-committee for review.

**Comment by philbucher** [comment]
@KratosMultiphysics/technical-committee I am not yet fully sure if this is needed, I wanted to look at it in more detail with @jcotela for doing #2993

File: kratos/includes/node.h [hunk 1]
```diff
@@ -757,42 +757,60 @@ class Node : public Point,  public IndexedObject, public Flags
...
                 return *it;
             }
         }
 
         // Otherwise do a find
-        it = mDofs.find(rDofVariable.Key());
-        if ( it!= mDofs.end() )
-        {
-            return *it;
+        return GetDof(rDofVariable);
```
Comment (RiccardoRossi) [inline kratos/includes/node.h]: here you must access again to begin and end, so it is slower... i would leave the implementaion you had before...

File: kratos/includes/node.h [hunk 2]
```diff
@@ -757,42 +757,60 @@ class Node : public Point,  public IndexedObject, public Flags
...
+            if( (it)->GetVariable() == rDofVariable) {
+                return *it;
+            }
         }
 
-        std::stringstream buffer;
-        buffer << "Not existant DOF in node #" << Id() << " for variable : " << rDofVariable.Name();
-        KRATOS_ERROR <<  buffer.str() << std::endl;
+        // Otherwise do a find
+        return GetDof(rDofVariable);
```
Comment (RiccardoRossi) [inline kratos/includes/node.h]: same comment...i would leave the version of before as it may be slightly more efficient and this is a very hot function

File: kratos/includes/node.h [hunk 3]
```diff
@@ -757,42 +757,60 @@ class Node : public Point,  public IndexedObject, public Flags
...
-        if ( it!= mDofs.end() )
-        {
+        if ( it!= mDofs.end() ) {
             return *it;
         }
 
-        std::stringstream buffer;
-        buffer << "Not existant DOF in node #" << Id() << " for variable : " << rDofVariable.Name();
-        KRATOS_ERROR <<  buffer.str() << std::endl;
+        KRATOS_ERROR <<  "Not existant DOF in node #" << Id() << " for variable : "
```
Comment (RiccardoRossi) [inline kratos/includes/node.h]: just wondering if this should not only be thrown in debug for efficiency reasons (i appreciate it would be a behaviour change and does not belong in this PR)

File: kratos/includes/node.h (hunk 1 above)
Comment (pooyan-dadvand) [reply kratos/includes/node.h]: I'm not sure if it is slower... std vector has begin and end pointers as member so optimizer would substitute it. Nevertheless you may not trust to optimizer which is understandable....

File: kratos/includes/node.h (hunk 3 above)
Comment (pooyan-dadvand) [reply kratos/includes/node.h]: If it is not an error it would not arrive to this statement. So I don't see it as a performance problem

File: kratos/includes/node.h [hunk 4]
```diff
@@ -801,34 +819,46 @@ class Node : public Point,  public IndexedObject, public Flags
         return mDofs;
     }
 
+    /** returns all of the Dofs  */
+    DofsContainerType& GetDofs() const
```
Comment (pooyan-dadvand) [inline kratos/includes/node.h]: why we need to give this access?

File: kratos/includes/node.h [hunk 5]
```diff
@@ -757,42 +757,60 @@ class Node : public Point,  public IndexedObject, public Flags
         typename DofsContainerType::iterator it_begin = mDofs.begin();
```
Comment (pooyan-dadvand) [inline kratos/includes/node.h]: I imagine that we don't need the non const version of this one. This should be used only for fast access but not for filling dof

**Review by pooyan-dadvand — COMMENTED** [review]
Is it possible to avoid all non const version of the GetDof and rely on pAddDof instead of them? Is it used in  many places?

File: kratos/includes/node.h (hunk 4 above)
Comment (philbucher) [reply kratos/includes/node.h]: I don't know, this was not added by me

**Comment by philbucher** [comment]
I updated the functions, now all `GetDof`-function are `const`

**Comment by pooyan-dadvand** [comment]
Are applications compiling with this change to const?

**Comment by philbucher** [comment]
@pooyan-dadvand I compiled the following apps sucessfully:
~~~
                STRUCTURAL_MECHANICS_APPLICATION - ON
                CONTACT_STRUCTURAL_MECHANICS_APPLICATION - ON
                SOLID_MECHANICS_APPLICATION - ON
                CONTACT_MECHANICS_APPLICATION - ON
                CONSTITUTIVE_MODELS_APPLICATION - ON
                FLUID_DYNAMICS_APPLICATION - ON
                MESH_MOVING_APPLICATION - ON
                FSI_APPLICATION - ON
                EMPIRE_APPLICATION - ON
                PFEM_SOLID_MECHANICS_APPLICATION - ON
                DEM_APPLICATION - ON
                EXTERNAL_SOLVERS_APPLICATION - ON
                TRILINOS_APPLICATION - ON
                METIS_APPLICATION - ON
                MESHING_APPLICATION - ON
                SHAPE_OPTIMIZATION_APPLICATION - ON
                MAPPING_APPLICATION - ON
                EIGEN_SOLVERS_APPLICATION - ON
                HDF5_APPLICATION - ON
                DELAUNAY_MESHING_APPLICATION - ON
~~~

These are the majority of Apps I would say

Let me know if I should also compile other apps

**Comment by pooyan-dadvand** [comment]
Perfect! Since it changes the behaviour I would wait for @KratosMultiphysics/technical-committee to approve it.

**Comment by philbucher** [comment]
ping

**Review by pooyan-dadvand — DISMISSED** [review]
To be documented, in @KratosMultiphysics/technical-committee we see the return of no constant reference from GetDof const and no constant pointer from pGetDof somehow non-standard and less intuitive. Nevertheless we can postpone the having two version of them with the constant one returning const reference/pointer to future.

**Comment by philbucher** [comment]
@KratosMultiphysics/technical-committee sorry I forgot to add `const` to the return type
I did it now bcs I think it is cleaner this way. Sorry for the inconvenience
I compiled almost the entire Kratos and ran the tests, the modifications that were necessary are minor

**Comment by philbucher** [comment]
now only `node.h` is changed, the code compiles
Also the return types are now `const`

**Comment by RiccardoRossi** [comment]
I agree it is nicer...however i don't see how this can work.

in the update we need to have a doflist and we use it for update. how can that work if the dofs are const?

**Comment by philbucher** [comment]
> I agree it is nicer...however i don't see how this can work.

@RiccardoRossi just trust the compiler on that one ;)

on a more serious note:
The Dof is not const, the pointer to the dof is!
> in the update we need to have a doflist and we use it for update. how can that work if the dofs are const?

**Comment by philbucher** [comment]
> I agree it is nicer...however i don't see how this can work.

@RiccardoRossi just trust the compiler on that one ;)

on a more serious note:
The Dof is not const, the pointer to the dof is!
> in the update we need to have a doflist and we use it for update. how can that work if the dofs are const?

**Comment by philbucher** [comment]
ping

**Comment by philbucher** [comment]
ping @KratosMultiphysics/technical-committee 
It would be nice to merge this then  can do the necessary changes for #2993

**Comment by RiccardoRossi** [comment]
+1 for merging this

**Comment by philbucher** [comment]
well someone has to approve it then :)

**Review by pooyan-dadvand — APPROVED** [review]
:+1:

**Comment by philbucher** [comment]
thx :+1:
